src/jit/hwintrinsic.cpp

```diff
@@ -138,4 +173,556 @@ CORINFO_CLASS_HANDLE Compiler::gtGetStructHandleForHWSIMD(var_types simdType, va
+        CORINFO_ARG_LIST_HANDLE argList  = sig->args;
+        CORINFO_CLASS_HANDLE    argClass;
+        var_types               argType = TYP_UNKNOWN;
+
+        assert(numArgs >= 0);
+        if ((HWIntrinsicInfo::lookupIns(intrinsic, baseType) == INS_invalid) ||
+            ((simdSize != 8) && (simdSize != 16) && (simdSize != 32)))
+        {
+            assert(!"Unexpected HW Intrinsic");
+            return nullptr;
+        }
+
+        GenTreeHWIntrinsic* retNode = nullptr;
+        GenTree*            op1     = nullptr;
+        GenTree*            op2     = nullptr;
+
+        switch (numArgs)
+        {
+            case 0:
+            {
+                retNode = gtNewSimdHWIntrinsicNode(retType, intrinsic, baseType, simdSize);
+                break;
+            }
+
+            case 1:
+            {
+                argType = JITtype2varType(strip(info.compCompHnd->getArgType(sig, argList, &argClass)));
+                op1     = getArgForHWIntrinsic(argType, argClass);
+                if ((category == HW_Category_MemoryLoad) && op1->OperIs(GT_CAST))
+                {
+                    // Although the API specifies a pointer, if what we have is a BYREF, that's what
+                    // we really want, so throw away the cast.
+                    if (op1->gtGetOp1()->TypeGet() == TYP_BYREF)
+                    {
+                        op1 = op1->gtGetOp1();
+                    }
+                }
+                retNode = gtNewSimdHWIntrinsicNode(retType, op1, intrinsic, baseType, simdSize);
+                break;
+            }
+
+            case 2:
+            {
+                argType = JITtype2varType(
+                    strip(info.compCompHnd->getArgType(sig, info.compCompHnd->getArgNext(argList), &argClass)));
+                op2 = getArgForHWIntrinsic(argType, argClass);
+
+                op2 = addRangeCheckIfNeeded(intrinsic, op2, mustExpand);
+
+                argType = JITtype2varType(strip(info.compCompHnd->getArgType(sig, argList, &argClass)));
+                op1     = getArgForHWIntrinsic(argType, argClass);
+
+                retNode = gtNewSimdHWIntrinsicNode(retType, op1, op2, intrinsic, baseType, simdSize);
+                break;
+            }
+
+            case 3:
+            {
+                CORINFO_ARG_LIST_HANDLE arg2 = info.compCompHnd->getArgNext(argList);
+                CORINFO_ARG_LIST_HANDLE arg3 = info.compCompHnd->getArgNext(arg2);
+
+                argType      = JITtype2varType(strip(info.compCompHnd->getArgType(sig, arg3, &argClass)));
+                GenTree* op3 = getArgForHWIntrinsic(argType, argClass);
+
+                op3 = addRangeCheckIfNeeded(intrinsic, op3, mustExpand);
+
+                argType = JITtype2varType(strip(info.compCompHnd->getArgType(sig, arg2, &argClass)));
+                op2     = getArgForHWIntrinsic(argType, argClass);
+
+#ifdef _TARGET_XARCH_
+                var_types op2Type = TYP_UNDEF;
```

Comment:
I don't see any reason either. I've merged them.

Reply:
It looks like it was split because `gtIndexBaseType` for the `retNode` needs to be set based on the `argClass` of `op2`. However, you have to get `op1` before the `retNode` can be created and getting `op1` mutates `argClass`.

I solved the issue by stashing the `op2ArgClass` in a local.